### DISCUSSION TITLE
Asynchronous tasks support

### DISCUSSION THREAD

**Discussion by meowfishorg** [celery/celery#9049] [Q&A]
I'm writing a FastAPI app (asynchronously) and want to use a task queue for offloading CPU intensive (and long running) tasks. Celery doesn't provide asynchronous support, although I can write the tasks with synchronous code, it is not convenient. 

Is there anyway I can directly use an async function for a task? If not, are there other open source alternatives that can do the job? Please recommend some **popular** ones (something like [taskiq](https://github.com/taskiq-python/taskiq)). Thanks!

Or are there better ways to do this, like writing your own?

**Comment by lucasgadams** [comment]
Celery maintainers seem to have an ideological stance against asyncio. This question gets asked all the time, there is a lot of demand for asyncio support, and it has been promised for many years. However I don't know what the status is or whether it will ever be supported. At the end of the day it is open source so can't complain too much I guess. I really think someone should just create a new framework that is modern and doesn't have the tech debt of Celery. There are some ways that you can make it work today, we use `celery_aio_pool`, but it is pretty jank and is more or less unmaintained. It has sufficed so far, but I don't think the situation is stable and I am very interested in migrating away.

**Reply by auvipy** [reply]
> Celery maintainers seem to have an ideological stance against asyncio.

please do not spread anything superficial. we had time and fund shortage. also I was sick from the last december, and recovering slowly from mid april. we will do everything needed according to our available time and fund, and also managing our personal life. if you have super urgent need, please feel free to fund my time to work on this.

**Comment by rkrell** [comment]
To be honest, I don't understand the "ideology" against asyncio. Many frameworks show they are able to handle both approaches.

We have had the same requirement to launch asynchronous tasks with Celery and found a fully working solution (using Celery 5.6. at the moment), please find an inspiration below and tell whether this works fine also for you:

```python
import asyncio
from inspect import isawaitable

import celery


class AsyncCeleryTask(celery.Task):
    abstract = True

    async def _run(self, *args, **kwargs):
        result = celery.Task.__call__(self, *args, **kwargs)
        if isawaitable(result):
            return await result
        return result

    def __call__(self, *args, **kwargs):
        return asyncio.run(self._run(*args, **kwargs))


class AsyncCelerySingleton(celery.Celery):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, task_cls=AsyncCeleryTask, **kwargs)


def create_celery():
    celery_app = AsyncCelerySingleton()
    celery_app.conf.update(
        broker_url=CONFIG.celery.broker_url,
        result_backend=CONFIG.celery.result_backend,
        broker_connection_retry_on_startup=CONFIG.celery.broker_connection_retry_on_startup,
        worker_redirect_stdouts_level=CONFIG.celery.worker_redirect_stdouts_level,
        worker_redirect_stdouts=CONFIG.celery.worker_redirect_stdouts,
        worker_hijack_root_logger=CONFIG.celery.worker_hijack_root_logger,
        task_serializer="pickle",
        accept_content=["pickle", "json"],
        result_serializer="pickle",
        task_routes = {
            'backend.celery_app.tasks.do_something.tasks.*': {'queue': 'important_work'}
        }
    )
    celery_app.autodiscover_tasks(packages=['backend.celery_app.tasks'])

    return celery_app
```

**Comment by ebram96** [comment]
Some workaround:

```python3

def async_task_runner(func):
    """
    A decorator for Celery tasks that runs the decorated async function
    within an asyncio event loop.
    """
    @wraps(func)
    def wrapper(*args, **kwargs) :
        return asyncio.run(func(*args, **kwargs))
    return wrapper


@async_task_runner
async def my_task(self):
    await some_async_work()

```
